crates/ecs/src/systems/mod.rs

```diff
@@ -607,61 +802,110 @@ macro_rules! impl_system_parameter_function {
+
+                            #[allow(trivial_casts)]
+                            $(let [<borrowed_pointer_$parameter>]: *mut _ = &mut borrowed.0.$parameter;)*
+
+                            // SAFETY: When querying on multiple threads, all segments need a mutable reference to borrowed.
+                            // Borrowed is never mutated except when the Write system parameter is used.
+                            // To ensure safety, intra-parallelization is disabled if any system parameter needs write access.
+                            // This check is done using the "support_parallelization" method on the Query system parameter.
+                            for _ in 0..segment_count {
+                                segments.push((
+                                    ($([<P$parameter>]::split_borrowed_data(unsafe { &mut *[<borrowed_pointer_$parameter>] }, FixedSegment::Single),)*),
+                                    borrowed.1,         // world
+                                    borrowed.2.clone(), // archetypes
+                                ));
+                            }
 
-                    let universe = self.world.get_archetype_indices(&base_signature);
+                            segments
+                        }
+                    }
+                }
 
-                    let archetypes_indices: Vec<_> = intersection_of_multiple_sets(&[
-                        universe.clone(),
-                        $([<P$parameter>]::filter(&universe, self.world),)*
-                    ])
-                    .into_iter()
-                    .collect();
+                unsafe fn fetch_parameter(
+                    segment: &mut Self::SegmentData<'_>,
+                ) -> Option<Option<Self>> {
+                    $([<P$parameter>]::reset_iterator_state(&mut segment.0.$parameter[0]);)*
+
+                    Some(Some(Self {
+                        phantom: PhantomData::default(),
+                        segments: std::mem::transmute(&mut segment.0), // the segment reference will always be valid in the system body
+                        world: std::mem::transmute(segment.1), // the world reference will always be valid in the system body
+                        archetypes: segment.2.clone(),
+                        iterate_over_entities: $([<P$parameter>]::iterates_over_entities())||*,
+                        iterated_once: false,
+                    }))
+                }
 
-                    let borrowed = (
-                        $([<P$parameter>]::borrow(self.world, &archetypes_indices)?,)*
-                    );
+                fn component_accesses() -> Vec<ComponentAccessDescriptor> {
+                    <($([<P$parameter>],)*) as SystemParameters>::component_accesses()
+                }
 
-                    let iterate_over_entities = $([<P$parameter>]::iterates_over_entities() ||)* false;
+                fn iterates_over_entities() -> bool {
+                    false
+                }
```

Comment:
I see the confusion. The `ìterate_over_entites` function tells whether a system needs to iterate over the entities when fetching parameters. See this example:

```rust
fn system_a(a: Read<A>, query: Query<(Read<A>,)>);
fn system_b(query: Query<(Read<A>,)>);
```

When running `system_a`, it will iterate over all entities with component `A`.
But with `system_b`, this system will not iterate and only run once per tick.

Do you have any suggestion on what `ìterate_over_entites` could be called instead?

Also, `$([<P$parameter>]::iterates_over_entities() ||)* false` is also completely unrelated: This function was not changed, it was just moved (previously on line number 530).

Reply:
> Also, `$([<P$parameter>]::iterates_over_entities() ||)* false` is also completely unrelated: This function was not changed, it was just moved (previously on line number 530).

Oh, that was confusing lol

> The `ìterate_over_entities function tells whether a system needs to iterate over the entities when fetching parameters.

Why specifically when fetching parameters? In the example, it seems like both `system_a` and `system_b` would need to fetch parameters for multiple entities, the difference between them being how many times the system itself is called. 

> Do you have any suggestion on what `iterates_over_entities` could be called instead?

Maybe `contains_multiple_entities`, since it's a system parameter which contains data from multiple entities while the other system parameters only contain data from a single entity. (But you'd have to negate the implementations and uses of it, in this case.) I think that would make it more understandable, since it more clearly describes what aspect of the system parameter that it is giving information about, rather than instructing an executor about how to iterate over it.

Reply:
> Why specifically when fetching parameters? In the example, it seems like both system_a and system_b would need to fetch parameters for multiple entities, the difference between them being how many times the system itself is called.

`fetch_parameter` for `Read`/`Write` will return `Some(Component)` and then `None` when it reaches the end of the last archetype.
`fetch_parameter` for `Query` will always return `Some(Query)` and the actual parameter fetching happens in `fetch_parameter` for the parameters inside `Query` when iterating over it.
If `fetch_parameter` eventually will return `None`, then `ìterate_over_entites` is true.

> Maybe contains_multiple_entities, since it's a system parameter which contains data from multiple entities while the other system parameters only contain data from a single entity.

`contains_multiple_entities` would only make sense for `Query`, `iterate_over_entites` is also false for filters and will be false for resources and command buffers.

Reply:
> `fetch_parameter` for `Query` will always return `Some(Query)` and the actual parameter fetching happens in `fetch_parameter` for the parameters inside `Query` when iterating over it.

Ah, I see. So  `Read`/`Write` parameters are eagerly fetched, while `Query` parameters are lazily fetched?

> If `fetch_parameter` eventually will return `None`, then `iterates_over_entities` is true.

> `contains_multiple_entities` would only make sense for `Query`, `iterates_over_entities` is also false for filters and will be false for resources and command buffers.

Then it sounds like the function represents whether the parameter contains a finite amount of data. Ordinary `Read`/`Write` parameters only contain data once for each entity, which is why they eventually return `None`, while special parameters such as `Query` and filters always return something and is there for not finite. 

So maybe it could be renamed to `contains_finite_data` or similar? Or maybe that's even more vague, heh. What do you think?

(As an aside: the whole `None`/`Some`-business and pseudo-iterator in `Read` and `Write` is still not ideal. Now that you seem to have figured out a way to implement `Iterator` for `Query`, do you think we could make all `SystemParameter`s into iterators?)